### ISSUE TITLE
Calling DrawImage with top image that overhangs bottom image in Y direction fails with exception

### ISSUE THREAD

**Issue by aceoft** [SixLabors/ImageSharp#2603]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have verified that I am running the latest version of ImageSharp
- [X] I have verified if the problem exist in both `DEBUG` and `RELEASE` mode
- [X] I have searched [open](https://github.com/SixLabors/ImageSharp/issues) and [closed](https://github.com/SixLabors/ImageSharp/issues?q=is%3Aissue+is%3Aclosed) issues to ensure it has not already been reported

### ImageSharp version

3.1

### Other ImageSharp packages and versions

SixLabors.ImageSharp.Drawing 2.0.1

### Environment (Operating system, version and so on)

Win10 22H2

### .NET Framework version

7

### Description

After upgrading the ImageSharp nuget package today, several of my unit tests started failing. They are all related to overlaying images that hang over the bottom edge of the bottom image.

The exception being thrown is:

`System.AggregateException: One or more errors occurred. (Specified argument was out of the range of valid values. (Parameter 'DangerousGetRowSpan(100). Y was out of range. Height=100')) ---> System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'DangerousGetRowSpan(100). Y was out of range. Height=100')`

It only seems to happen when a `Y` value is larger than the height of the bottom image.

I'm just getting started with ImageSharp and have been impressed by the amount of thought and hard work that has been put into the library. Thank you, and hopefully I provided everything we need here.

### Steps to Reproduce

Test method to reproduce the exception:

```csharp
public void TestOverlayImageOffEdge()
{
	using var bottom = new Image<Rgb24>(100, 100, Color.Black);
	using var top = new Image<Rgb24>(100, 100, Color.White);

	// OK (expected)
	bottom.Mutate(i => i.DrawImage(top, new Point(0, 0), 1f));

	// OK (expected, overhangs X direction by 1 pixel)
	bottom.Mutate(i => i.DrawImage(top, new Point(1, 0), 1f));

	// throws System.AggregateException
	bottom.Mutate(i => i.DrawImage(top, new Point(0, 1), 1f));
}
```

I had a quick look at the source code and wondered if providing a source rectangle would help -- it did. The below line appears to be a workaround.

```csharp
// OK
bottom.Mutate(i => i.DrawImage(top, new Point(0, 1), new Rectangle(0, 0, 100, 99), 1f));
```

**Comment by aceoft** [comment]
It appears that more than one behavior was changed in this area. I am also having trouble with overlaying images with the origin point being outside the dimensions of the bottom image.

For example, an origin of -100, -100 appears to be getting reset to 0, 0 before overlaying now.

**Comment by JimBobSquarePants** [comment]
Yeah, lots of changes. 

https://github.com/SixLabors/ImageSharp/issues/2447

**Comment by aceoft** [comment]
Thank you for your response! I'm sorry, I wasn't clear enough in my last statement. I meant passing negative values for the `X` or `Y` values in the `location` (renamed to `backgroundLocation`) argument.

In 3.0.2, the top image would originate from that location correctly (you end up losing part of the left/top of the top image as intended). In 3.1, it doesn't seem to allow a `backgroundLocation` with negative values. It silently resets to `0, 0` and draws from there.

**Comment by tehsoul** [comment]
Just want to chime in that I see the exact same behavior of appearing to not respect negative origins and resetting to 0 after having bumped to 3.1.0.
I was going to investigate and create an issue if I got a few hours for it but you beat me to it with this great explanation!
I reverted to 3.0.2 for now

**Comment by JimBobSquarePants** [comment]
@tehsoul Please do not revert. That's the worst possible thing you could do!!!

The behavior in 3.0.2 is incorrect and you are basing your implementations on incorrect behavior. I will try to explain why.

Take for example the overlay image shared in issue https://github.com/SixLabors/ImageSharp/issues/2447

![image](https://github.com/SixLabors/ImageSharp/assets/385879/3c84e71b-768d-4cfd-bfa9-82652b3dc7ed)

If we wanted to render the following result that contains what you both would consider a negative offset off the overlay

![image](https://github.com/SixLabors/ImageSharp/assets/385879/f06de277-80fe-4515-8753-86a764eb486f)

you should write the following code:

```c#
using var src = Image.Load<Rgba32>("numbergrid.png");
using var dst = new Image<Rgba32>(100, 100, new Rgba32(0, 255, 255));

dst.Mutate(c => {
    c.DrawImage(src, new Point(64, 10), new Rectangle(32, 32, 32, 32), 1.0f);
});

dst.SaveAsPng("result.png");
```

As you can see from the code says to render the source image at position `[64,10]` taking from the source image a rectangle at position `[32,32]` and size `32x32`.

To consider this a negative offset in coordinate terms is incorrect. The rectangle parameter specifically represents the area of interest within the overlay bounds to draw.  

Ignoring negative values is a common way to normalize out of bounds properties hence why `Rectangle.Intersect` exists in here, System.Drawing, and many other libraries.

@aceoft What you think is the "intended behavior" is absolutely not intended at all. We explicitly want to avoid negative values in the overlay rectangle because they simply do not make sense. 

I will repeat this again. The behavior you saw in v3.0.2 was unintentional and incorrect. 

Does that make things clear?

**Comment by JimBobSquarePants** [comment]
I'll have a look at the original issue but there will be no reversion to the buggy behavior present in previous versions.

**Comment by tehsoul** [comment]
@JimBobSquarePants I agree with the logic in your example. 
However, here is a very simplified version of what I'm trying to achieve (screenrecording from manually doing the action I want to automate in Paint.NET):

![imagesharpregression](https://github.com/SixLabors/ImageSharp/assets/127198586/5404e7cc-52f6-439e-a900-f4ed2b0eab69)

--> basically my code starts with the top left point as input, which is where I'm drawing it on the destination layer:

```csharp
using var src = Image.Load<Rgba32>("numbergrid.png");
using var dst = new Image<Rgba32>(512, 512, Color.HotPink);

dst.Mutate(x=>x.DrawImage(src, new Point(-300, -150), 1.0f));

dst.SaveAsPng("result.png");
```

3.0.2 output:
![result](https://github.com/SixLabors/ImageSharp/assets/127198586/ade2e2b9-2988-4d33-8379-6a46f12cb770)

3.1.0 output (for my scenario, totally not what is intended):
![result](https://github.com/SixLabors/ImageSharp/assets/127198586/5636392a-0dc4-422a-8b33-fdd582b14928)

**Comment by tocsoft** [comment]
@JimBobSquarePants the difference is between passing a `Rectangle` and just a `Point`. Based on the images provided I would have to agree that behaviour is broken in the case of a `Point` rather than a `Rectangle`. For a `Point` I would expect a negative offset to work and show the bottom right of the overlay image, however when passing a `Rectangle` I do agree it should be the explicit region defined.

I.e. Passing a `Rectangle` implies a crop followed by a draw operation, but passing a `Point` implies a straight draw operation leaving only the part of the overlay that happens to intersect visible.

**Comment by tehsoul** [comment]
To add to this, this behavior is also weird and imo inconsistent in 3.1.0:

## Moving the image positively on the X axis so it overlaps destination - **WORKS**
```csharp
using var src = Image.Load<Rgba32>("numbergrid.png");
using var dst = new Image<Rgba32>(512, 512, Color.HotPink);

dst.Mutate(x => x.DrawImage(src, new Point(300, 0), 1.0f));

dst.SaveAsPng("result.png");

--> works as expected
```
![result](https://github.com/SixLabors/ImageSharp/assets/127198586/8216a785-7f98-4385-9d1d-809cc0cb5acf)

## Moving the image positively on the Y axis so it overlaps destination - **FAILS WITH EXCEPTION**
```csharp
using var src = Image.Load<Rgba32>("numbergrid.png");
using var dst = new Image<Rgba32>(512, 512, Color.HotPink);

dst.Mutate(x => x.DrawImage(src, new Point(300, 0), 1.0f));

dst.SaveAsPng("result.png");
```

Fails with error:

> System.AggregateException
One or more errors occurred. (Specified argument was out of the range of valid values. (Parameter 'DangerousGetRowSpan(537). Y was out of range. Height=512')) (Specified argument was out of the range of valid values. (Parameter 'DangerousGetRowSpan(623). Y was out of range. Height=512')) (Specified argument was out of the range of valid values. (Parameter 'DangerousGetRowSpan(580). Y was out of range. Height=512')) (Specified argument was out of the range of valid values. (Parameter 'DangerousGetRowSpan(512). Y was out of range. Height=512'))

--> this worked in 3.0.2 as expected:

![result](https://github.com/SixLabors/ImageSharp/assets/127198586/7e218306-af4a-4b21-916d-c31722e84b66)

**Comment by JimBobSquarePants** [comment]
@tocsoft Passing a negative point implies virtual pixels outside of the background image. I don’t think that is wise and is inconsistent with the rest of the library. It’s the same situation for our ImageBrush in Drawing.

@tehsoul that’s literally the raised issue. We should be clipping the vertical bounds

**Comment by tocsoft** [comment]
I would argue this is more inconstant with the other Drawing APIs. if you do the same operation with a vector I'm pretty sure it'll work as expected, i.e. drawing a circle with a negative offset will draw the bottom right portion of the circle not the top left.

**Comment by tehsoul** [comment]
For anyone running into this in the future, here is a naive custom extension method as a quick workaround for drawing to any point

@JimBobSquarePants this is basically the behavior from 3.0.2 when drawing to a point (which could be negative and/or cause a part of the foreground to be cut off on the Y axis if the target canvas didn't have enough vertical real estate)

```csharp
public static class DrawingExtensions
{
    /// <summary>
    /// Draws the given image together with the currently processing image by blending their pixels.
    /// </summary>
    /// <param name="source">The current image processing context.</param>
    /// <param name="foreground">The image to draw on the currently processing image.</param>
    /// <param name="backgroundLocation">The location on the currently processing image at which to draw. - can be negative or cause the result to flow over the source</param>
    /// <param name="opacity">The opacity of the image to draw. Must be between 0 and 1.</param>
    /// <returns>The <see cref="IImageProcessingContext"/>.</returns>
    public static IImageProcessingContext DrawImageToAPointRegressionProof(
        this IImageProcessingContext source,
        Image foreground,
        Point backgroundLocation,
        float opacity)
    {
        // revert to 0 for negative X and Y
        var newPoint = new Point(
            backgroundLocation.X < 0 ? 0 : backgroundLocation.X,
            backgroundLocation.Y < 0 ? 0 : backgroundLocation.Y);

        // adapt cropping rectangle accordingly
        var boundX = backgroundLocation.X < 0 ? -backgroundLocation.X : 0;
        var boundY = backgroundLocation.Y < 0 ? -backgroundLocation.Y : 0;
        var boundWidth = foreground.Bounds.Width - boundX;
        var boundHeight = foreground.Bounds.Height - boundY;

        // adjust (reduce) for output canvas if needed
        var canvasSize = source.GetCurrentSize();
        var remainingWidthOnCanvas = canvasSize.Width - newPoint.X;
        var remainingHeightOnCanvas = canvasSize.Height - newPoint.Y;
        
        boundWidth = Math.Min(boundWidth, remainingWidthOnCanvas);
        boundHeight = Math.Min(boundHeight, remainingHeightOnCanvas);

        var newBounds = new Rectangle(
            boundX,
            boundY,
            boundWidth,
            boundHeight
        );

        GraphicsOptions options = source.GetGraphicsOptions();
        return source.ApplyProcessor(new DrawImageProcessor(foreground, newPoint, newBounds, options.ColorBlendingMode, options.AlphaCompositionMode, opacity));
    }
}
```


Tests:

```csharp
[Fact]
public void MoveToLeftAndUp()
{
    using var src = Image.Load<Rgba32>("numbergrid.png");
    using var dst = new Image<Rgba32>(src.Size.Width + 50, src.Size.Height + 50, Color.HotPink);

    dst.Mutate(x => x.DrawImageToAPointRegressionProof(src, new Point(-300, -150), 1.0f));

    dst.SaveAsPng("result1.png");
}
```
![result1](https://github.com/SixLabors/ImageSharp/assets/127198586/86282b8f-d020-4c82-a6ac-d8e88f994041)

```csharp
[Fact]
public void MoveToRightAndDown()
{
    using var src = Image.Load<Rgba32>("numbergrid.png");
    using var dst = new Image<Rgba32>(src.Size.Width + 50, src.Size.Height + 50, Color.HotPink);

    dst.Mutate(x => x.DrawImageToAPointRegressionProof(src, new Point(300, 150), 1.0f));

    dst.SaveAsPng("result2.png");
}
```
![result2](https://github.com/SixLabors/ImageSharp/assets/127198586/35a8e7b1-544c-4f2f-b6d9-7f2e55913ae4)

```csharp
[Fact]
public void MoveToRight()
{
    using var src = Image.Load<Rgba32>("numbergrid.png");
    using var dst = new Image<Rgba32>(src.Size.Width + 50, src.Size.Height + 50, Color.HotPink);

    dst.Mutate(x => x.DrawImageToAPointRegressionProof(src, new Point(300, 0), 1.0f));

    dst.SaveAsPng("result3.png");
}
```
![result3](https://github.com/SixLabors/ImageSharp/assets/127198586/db020b0a-5079-49fe-986e-d664b7a652b6)

```csharp
[Fact]
public void MoveDown()
{
    using var src = Image.Load<Rgba32>("numbergrid.png");
    using var dst = new Image<Rgba32>(src.Size.Width + 50, src.Size.Height + 50, Color.HotPink);

    dst.Mutate(x => x.DrawImageToAPointRegressionProof(src, new Point(0, 150), 1.0f));

    dst.SaveAsPng("result4.png");
}
```
![result4](https://github.com/SixLabors/ImageSharp/assets/127198586/5ea35209-a12d-45ad-bdac-976625ecf322)

```csharp
[Fact]
public void MoveToMiddle()
{
    using var src = Image.Load<Rgba32>("numbergrid.png");
    using var dst = new Image<Rgba32>(src.Size.Width + 50, src.Size.Height + 50, Color.HotPink);

    dst.Mutate(x => x.DrawImageToAPointRegressionProof(src, new Point(20, 20), 1.0f));

    dst.SaveAsPng("result5.png");
}
```
![result5](https://github.com/SixLabors/ImageSharp/assets/127198586/c93c23ca-0c42-49cb-8ae8-fc77a224046a)

```csharp
[Fact]
public void MoveToLeftAndUpOnASmallerCanvas()
{
    using var src = Image.Load<Rgba32>("numbergrid.png");
    using var dst = new Image<Rgba32>(src.Size.Width -50, src.Size.Height - 50, Color.HotPink);

    dst.Mutate(x => x.DrawImageToAPointRegressionProof(src, new Point(-20, -20), 1.0f));

    dst.SaveAsPng("result6.png");
}
```
![result6](https://github.com/SixLabors/ImageSharp/assets/127198586/c5d270c1-8778-40e8-b6a6-6970c669b571)

**Comment by tehsoul** [comment]
@JimBobSquarePants 
For what it's worth - I'm just trying to be thorough and convey the issue I'm talking about as clearly as I can.
Sorry!

**Comment by tocsoft** [comment]
fix has been merged

**Comment by Visual-Vincent** [comment]
It's a bummer that I didn't notice this issue earlier, as I might've been able to assist. Happy to see that it was resolved in such a swift manner, though!

Back when I wrote #2447 I was working on a solution to the issue myself and did manage to come up with one that I backported to 2.1.x as part of a WIP update to an older project of mine. Sadly due to a busy schedule I never got around to making a PR before @JimBobSquarePants had already prepared one, which is a bit unfortunate as based on my testing it seems I had also unknowingly solved this (future) issue.

My implementation was committed here for anyone interested, though the code is not as clean and neat as the official solution:
Visual-Vincent/DoomWriter@d521accb01b34b380a5daebd25fcba17a87c4a7e
